drivers/dfplayer: New driver for the DFPlayer Mini MP3 player

maribu opened the pull request:
### Contribution description

This PR adds a driver for the UART controlled MP3 player module sold as "DFplayer Mini" for about 2€ and less. The shell command `dfplayer` is provided as well (when both modules `shell_commands` and `dfplayer` are used). The new test application in `tests/driver_dfplayer` provides exactly this.

### Testing procedure

1. Get an SD card sized 32GB or less. (Those come with a DOS partition table with exactly one partition spanning the whole storage and formatted as FAT32. This is what the device needs, in case you manually changed the partition table or the file system.)
     - A USB stick works fine as well, but is trick to connect
2. Store some MP3 files on it using any (or all) of the following naming schemes:
    1. Files with three digit filenames (leading zeors) in two digit folder names (`"01/001.mp3"` to `"99/255.mp3"`)
    2. Files with four digit filenames stored in the folder MP3 (`"MP3/0001.mp3"` to `"MP3/9999.mp3"`)
    3. Files with four digit filenames stored in the folder ADVERT (`"ADVERT/0001.mp3"` to `"ADVERT/9999.mp3"`)
3. Plug the SD card into the DFPlayer Mini, connect it via UART (3.3V logic level) to your board
4. Connect the DFPlayer Mini to a speaker using either
    1. The SPK+ and SPK- pins to directly drive a passive speaker (mono) without the need of an amplifier. (An internal 3W amplifier is used.)
    2. The ADC_L, ADC_R and GND pins to directly drive a headphone, to directly connect to the AUX input via an audio Jack of your home entertainment system, or two to (stereo) speakers using a 2-channel amplifier
5. Compile, flash, and run `tests/driver_dfplayer` and use the `dfplayer` shell command to play music.

### Issues/PRs references

Supersedes https://github.com/RIOT-OS/RIOT/pull/12271

benpicco commented on drivers/dfplayer/dfplayer_internal.c:
```diff
@@ -0,0 +1,488 @@
+        uart_write(dev->uart, frame, sizeof(frame));
+
+        /* wait for reply from ISR */
+        dev->sync = thread_getpid();
+        thread_sleep();
+        dev->sync = KERNEL_PID_UNDEF;
```
That's an interesting way to do that.
I would have just called `mutex_lock()` (`xtimer_mutex_lock_timeout()`) here and unlocked it from the interrupt.

benpicco commented on drivers/dfplayer/dfplayer_internal.c:
```diff
@@ -0,0 +1,488 @@
+
+    /* Await notification the DFPlayer Mini is available */
+    /* set timeout timer */
+    xtimer_t timer = {
+        .callback = _timeout_cb,
+        .arg = dev
+    };
+    xtimer_set(&timer, DFPLAYER_BOOTUP_TIME_MS * US_PER_MS);
+    dev->sync = thread_getpid();
+    thread_sleep();
+    dev->sync = KERNEL_PID_UNDEF;
+
+    /* In case no timeout occurred, stop timer */
+    xtimer_remove(&timer);
```
I think `xtimer_mutex_lock_timeout()` would be a lot shorter :wink:

maribu replied:
Done

maribu commented on drivers/dfplayer/dfplayer_internal.c:
```diff
@@ -0,0 +1,488 @@
+#include "debug.h"
+
+/**
+ * @brief   Initial value of the FCS
```
Frame Check Sequence

benpicco replied:
Better spell it out

maribu replied:
Done

maribu commented:
> I guess that folder structure is given by the firmware on the player.

Indeed. My two flavors of the DFPlayer thankfully ignored any trailing chars after the number, so I got away with "01 Some Artist - Some Album/001 Some Title Name.mp4" instead of "01/001.mp3". However, the "datasheets" said nothing about this behaviour. On the other hand, the two datasheets I could find were extremely vague, often misleading, severely lacking important information, and sometimes just plain wrong. It could be that indeed every device with every firmware version handles this the same way. But I didn't dare to document this, if someone is unlucky and one's device failed with additional characters in the file  or folder name.

benpicco commented:
Want to give this a rebase so I can give this a try?

benpicco commented on tests/driver_dfplayer/Makefile:
```diff
@@ -0,0 +1,11 @@
+
+include $(RIOTBASE)/Makefile.include
+
+CFLAGS += '-DDFPLAYER_PARAM_BUSY_PIN=$(BUSY_PIN)'
```
This will cause problems if `DFPLAYER_PARAM_BUSY_PIN` is already defined in `board.h`.

maribu replied:
True. I'll drop the definition of `BUSY_PIN` and only extent `CFLAGS` if `BUSY_PIN` is defined.

benpicco commented:
I finally hooked up the DFPlayer to one of my `weact-f411ce`.

First I tried without modifying any filenames:

```
├── 01 Demon Days
│   ├── 01 Intro.mp3
│   ├── 02 Last Living Souls.mp3
│   ├── 03 Kids With Guns.mp3
│   ├── 04 O Green World.mp3
│   ├── 05 Dirty Harry.mp3
│   ├── 06 Feel Good Inc..mp3
│   ├── 07 El Mañana.mp3
│   ├── 08 Every Planet We Reach Is Dead.mp3
│   ├── 09 November Has Come.mp3
│   ├── 10 All Alone.mp3
│   ├── 11 White Light.mp3
│   ├── 12 Dare.mp3
│   ├── 13 Fire Coming Out Of The Monkey's Head.mp3
│   ├── 14 Don't Get Lost In Heaven.mp3
│   └── 15 Demon Days.mp3
└── MP3
    ├── 01 Goof.mp3
    ├── 02 Sandjorda.mp3
    ├── 03 Widibf.mp3
    └── 04 Fuayfsilfm.mp3

2 directories, 19 files
```

When I run `dfplayer play` the first file from the `mp3` folder is played.

I get

```
2020-06-12 18:15:17,502 #  dfplayer
2020-06-12 18:15:17,624 # Software version: 8
2020-06-12 18:15:17,624 # State: playing
2020-06-12 18:15:17,748 # Volume: 15
2020-06-12 18:15:17,749 # Selected file numbers (in file system):
2020-06-12 18:15:17,870 #   - SD card: 1
2020-06-12 18:15:17,870 # Current track: 0/0
2020-06-12 18:15:17,994 # Playback mode: normal
2020-06-12 18:15:17,994 # USB: Not present
2020-06-12 18:15:18,115 # SD card: 19
2020-06-12 18:15:18,116 # Flash: Not present
```

Can't play any other track though:

```
2020-06-12 18:14:54,638 #  dfplayer mp3 2
2020-06-12 18:14:54,860 # Error: No such file or directory
```

So I renamed my files

```
├── 01 Demon Days
│   ├── 001 Intro.mp3
│   ├── 002 Last Living Souls.mp3
│   ├── 003 Kids With Guns.mp3
│   ├── 004 O Green World.mp3
│   ├── 005 Dirty Harry.mp3
│   ├── 006 Feel Good Inc..mp3
│   ├── 007 El Mañana.mp3
│   ├── 008 Every Planet We Reach Is Dead.mp3
│   ├── 009 November Has Come.mp3
│   ├── 010 All Alone.mp3
│   ├── 011 White Light.mp3
│   ├── 012 Dare.mp3
│   ├── 013 Fire Coming Out Of The Monkey's Head.mp3
│   ├── 014 Don't Get Lost In Heaven.mp3
│   └── 015 Demon Days.mp3
└── MP3
    ├── 0001 Goof.mp3
    ├── 0002 Sandjorda.mp3
    ├── 0003 Widibf.mp3
    └── 0004 Fuayfsilfm.mp3

2 directories, 19 files
```

I can play files from the `MP3` folder

```
2020-06-12 18:30:16,889 #  dfplayer mp3 2
> dfplayer track
2020-06-12 18:30:38,977 #  dfplayer track
2020-06-12 18:30:38,978 # Selected file numbers (in file system):
2020-06-12 18:30:39,098 #   - SD card: 3
2020-06-12 18:30:39,099 # Current track: 2
```

But not from the `01 Demon Days` folder:

```
2020-06-12 18:31:10,088 #  dfplayer file 1 1
2020-06-12 18:31:10,310 # Error: No such file or directory
```

Well, let's rename that to just `01`:

```
2020-06-12 18:31:41,072 # dfplayer file 1 1
2020-06-12 18:32:17,071 #  dfplayer track
2020-06-12 18:32:17,073 # Selected file numbers (in file system):
2020-06-12 18:32:17,193 #   - SD card: 14
2020-06-12 18:32:17,193 # Current track: 1/1
```

Works!

But even with `dfplayer play`, playback stops after the first song.

```
2020-06-12 18:33:17,486 #  dfplayer play
> 2020-06-12 18:34:20,900 #  Playback of track 14 on device SD card has completed
2020-06-12 18:34:20,911 # Playback of track 14 on device SD card has completed
```

Shouldn't it play the whole folder?

maribu commented:
> When I run `dfplayer play` the first file from the `mp3` folder is played.

This is by only by chance the first file in that folder. Internally the device works on "tracks", with the track number depending on the order in the file system. If you would copy (on an empty SD Card) first file `mp3/0002.mp3` and then `mp3/0001.mp3`, the file `mp3/0002.mp3` would be track 1 and played instead :-/

maribu commented:
> Shouldn't it play the whole folder?

The DFPlayer does support continuing with the next "track" after finishing with one file natively. But the next "track" will be whatever file is next in the file system, not what is alphabetically next im the same folder.

There is one command that allows playing a specific folder. The order is still surprising, but at least files of other folders are skipped.

There is also one random playback mode that I'm sure I exposed in the shell command. This does continue with the next file after completing playbavk of a file. But that again ignores the folder structure and will also play "advertisements". That would be very inconvenient if the `ADVERT` folder is used for UI feedback sounds e.g. on volume changes or prev/next commads.  If not using "advertisements", this would however work just fine.

In the docu there is section on continuous playback that explains how to address this. Now with the event thread being merged, I could provide a submodule for continuous playback relying on that. I was reluctant to that previously, because allocating a thread just for selecting the next song upon completion of playback seemed like a huge waste.

benpicco commented on drivers/include/dfplayer.h:
```diff
@@ -0,0 +1,751 @@
+ * The MP3 files have to be named with four decimal digits (e.g. `"0001.mp3"` or
+ * `"1337.mp3") and must be placed in the folder `"MP3"`. The file numbers
+ * 1-9999 are valid. Playback of these files can be started using the function
+ * @ref dfplayer_play_from_mp3
```
Can you add a note that characters after the numbers are ignored?
I've seen this in other tutorials about the DFPlayer and it makes working with those files much easier.
(And that this is not true for folders)

benpicco commented on drivers/include/dfplayer.h:
```diff
@@ -0,0 +1,751 @@
+ * mono speaker. Alternatively a stereo output can be used that can be
+ * directly connected to headphones, but needs a dedicated amplifier for
+ * connecting to passive speakers.
+ *
```
Would be neat to have a pinout here
![DFPlayer Pinout](https://cdn.shopify.com/s/files/1/1509/1638/files/MP3_player_module_pinout_diagram.png)

maribu commented:
Rebased to solve merge conflict.

maribu commented:
> I guess there is no way to query what files/titles (by ID3 tag) are on the SD card?

At least not I'm aware of. I have seen micropython drivers doing a binary search for valid file names by trying to start playback and checking for success. With ~1 sec per test this still is ~8 secs per folder (in the `01/234.mp3` format). I assume no one would do this, if there would be a sane way of accessing file names or even tags.

maribu commented:
OK, I added two more things:

I changed the test application to use `event/thread` for the handlers and directly start the next song once playback is completed. I originally intended to add this as submodule; but now I think it is better let the application developer implement this and e.g. pick a different IPC mechanism instead. Having however an example implementation in the test will certainly not hurt.

This triggered a bug: Apparently my DFPlayer sends two playback completed messages, but previously I used stdio in IRQ context which blocked the board long enough to miss the second event. IMO the user of the driver should not worry about detecting duplicated events, so I extended the driver to perform the de-duplication instead.

benpicco commented on drivers/dfplayer/dfplayer_internal.c:
```diff
@@ -75,9 +75,11 @@ static void _handle_playback_completed(dfplayer_t *dev, dfplayer_source_t src)
 
-    if (dev->cb_done) {
+    uint32_t now_us = xtimer_now_usec();
+    if (dev->cb_done && (now_us - dev->last_event_us > US_PER_SEC)) {
```
Better add a comment that this is to prevent duplicate events - otherwise the casual reader might wonder what this is for and even attempt to remove it.

maribu replied:
Good idea. Done

benpicco commented on tests/driver_dfplayer/main.c:
```diff
@@ -85,7 +125,7 @@ int main(void)
     );
 
-    if (dfplayer_set_callbacks(dfplayer_get(0), _cb_done, _cb_src, NULL)) {
+    if (dfplayer_set_callbacks(dfplayer_get(0), _cb_done, _cb_src, dfplayer_get(0))) {
```
The pointer to the `dfplayer_t` struct is already passed to all callback functions already, no?

maribu replied:
It does indeed make more sense to just pass the device descriptor and safe the RAM of the user pointer. The user would be interested in anything other than the device descriptor anyway.

maribu replied:
Fixed

benpicco replied:
Having the ability to pass an arbitrary pointer to the callback is still useful - it's just not needed in this example.

maribu replied:
You can still do this e.g. using:

```C
typedef struct {
    dfplayer_t dfplayer;
    void *userptr;
} dfplayer_with_user_ptr;
```

And now using the `container_of()` macro will give you the user pointer. It has the disadvantage of no longer being able to use `auto_init()`, but it safes 4 byte for most use cases.

benpicco commented:
I can confirm that playback of serveral tracks in a row is now working!

Track numbers are a bit off though
```
> dfplayer mp3 1
> dfplayer mode
2020-06-25 22:29:11,995 # Playback mode: normal
> dfplayer state
2020-06-25 22:29:17,139 # State: playing
> dfplayer track
2020-06-25 22:29:19,891 # Selected file numbers (in file system):
2020-06-25 22:29:20,011 #   - SD card: 16
2020-06-25 22:29:20,012 # Current track: 1
2020-06-25 22:33:05,976 #  Playback of track 16 on device SD card has completed
2020-06-25 22:37:50,005 #  Playback of track 18 on device SD card has completed
> dfplayer track
2020-06-25 22:38:05,843 # Selected file numbers (in file system):
2020-06-25 22:38:05,964 #   - SD card: 17
2020-06-25 22:38:05,964 # Current track: 3
```

maribu commented:
> Track numbers are a bit off though

Note that "track" revers to the internal filesystem order... One would have to prepare an SD card at a very low level in order to actually make use of the track number.

Maybe I should rephrase the message. The function `dfplayer_get_track()` is your friend if you want to figure out what track in regard to the file naming convention is currently played. But that function only works if playback was started according to one of the two naming conventions, rather than e.g. shuffle playback of all files or repeat playback of a folder.

maribu commented:
The test now gives additionally the file name (or the leading part of it) out on playback completed, e.g.:

```
2020-06-25 23:24:37,220 #  Playback of track 30 on device SD card has completed
2020-06-25 23:24:37,222 # 0002.mp3: Playback completed
```

or 

```
2020-06-25 23:21:11,101 #  Playback of track 16 on device SD card has completed
2020-06-25 23:21:11,103 # 01/002.mp3: Playback completed
```

With both infos on screen, it will become obvious that the "track" is a pretty bogus unit. (I would rephrase it, but the datasheet names that number "track", so I sticked with it.)

benpicco commented on sys/shell/commands/sc_dfplayer.c:
```diff
@@ -0,0 +1,617 @@
+static void _print_error(int retval)
+{
+    print_str("Error: ");
+    print_str(strerror(-retval));
```
looks like this is not in the AVR libc.

maribu replied:
That makes sense. On a Harvard arch, this would result in a lot of error messages being allocated in RAM rather than in flash. Fixed.

maribu commented:
I extended the doc a bit and renamed `dfplayer_get_track()` to `dfplayer_get_played_file()`. Now `track` is consistently used for the bogus unit, wile `dfplayer_file_t` is used for the file according to the used naming scheme. That should be less confusing.

benpicco commented on tests/driver_dfplayer/Makefile:
```diff
@@ -10,3 +10,8 @@ include $(RIOTBASE)/Makefile.include
   CFLAGS += '-DDFPLAYER_PARAM_BUSY_PIN=$(BUSY_PIN)'
 endif
+
+ifneq (,$(filter arch_avr8,$(FEATURES_USED)))
+  # no strerror() in avrlibc; which makes a lot of sense on a Harvard machine
+  CFLAGS += -DDFPLAYER_NO_STRERROR
+endif
```
That should probably go into `Makefile.dep` instead.

aabadie replied:
I would prefer this in Makefile.include. Now that the order of inclusion is merged, let's try to do things right.

maribu replied:
Indeed. I first thought `strerror()` would be used in the `test`, not the shell command, at which point it made sense to put it only locally in the tests `Makefile`. I forgot to reconsider that decision

maribu replied:
> I would prefer this in Makefile.include. Now that the order of inclusion is merged, let's try to do things right.

:+1:. I will rebase (without squashing) to make local testing easier

benpicco commented:
Murdock found some issues

benpicco commented:
Murdock is not happy.

maribu commented:
May I squash?

benpicco commented:
Sure, squash away!

benpicco commented:
Travis is still not happy.

benpicco commented:
Travis is happy now - please squash!

maribu commented:
Thanks for the review!
